src/searcher/core.rs

```diff
@@ -0,0 +1,258 @@
+use crate::{
+    lines::{self, LineStep},
+    query_context::QueryContext,
+    searcher::{Config, Range, Searcher},
+    sink::{Sink, SinkContext, SinkContextKind, SinkFinish, SinkMatch},
+};
+
+#[derive(Debug)]
+pub struct Core<'s, S> {
+    config: &'s Config,
+    query_context: QueryContext,
+    searcher: &'s Searcher,
+    sink: S,
+    pos: usize,
+    absolute_byte_offset: u64,
+    line_number: Option<u64>,
+    last_line_counted: usize,
+    last_line_visited: usize,
+    after_context_left: usize,
+    has_sunk: bool,
+}
+
+impl<'s, S: Sink> Core<'s, S> {
+    pub fn new(searcher: &'s Searcher, query_context: QueryContext, sink: S) -> Core<'s, S> {
```

Comment:
Ok so basically a `QueryContext` gets passed mostly everywhere where a `Matcher` was getting passed before, and then in the appropriate place we "inline" use the `QueryContext` to get an iterator over tree-sitter query capture matches (so basically the searcher has tree-sitter-specific stuff hardcoded in)